doc/ip/next_hop_group_hld.md

```diff
@@ -0,0 +1,148 @@
+| 0.1 | 25/11/2020  |   Thomas Cappleman      | Initial version                            |
+
+
+# About this Manual
+This document provides general information about an enhancement to the internal APP_DB routing table to split next hop information out into its own table.
+
+Throughout this document a next hop group may refer to either a single next hop or a group of several next hops.
+
+# 1 Introduction
+Currently the route table within APP_DB contains all next hop information for a route embedded in that route's entry in the table.
+
+At high scale (particularly when handling millions of routes all routed over multiple next hops) this is inefficient both in terms of performance and occupancy.  A given route entry needs to be built up including all its next hop information, which gets sent to and stored in APP_DB.  Orchagent then needs to receive, parse and handle all the information in that route entry every time.
+
+A more efficient system would involve managing the next hop groups in use by the route table separately, and simply have the route table specify a reference to which next hop group to use.  Since at scale many routes will use the same next hop groups, this requires much smaller occupancy per route, and so more efficient building, transmission and parsing of per-route information.
+
+# 2 Requirement Overview
+## 2.1 Functional Requirements
+
+- Improve efficiency of storage of routing information in APP_DB.
+- Reduce time to reprogram routes when next hop information changes.
+
+## 2.2 Configuration and Management Requirements
+- The existing 'show ip route' and 'show ipv6 route' commands must be supported when using the new next hop group table.
+
+## 2.3 Scalability Requirements
+- Unchanged in requirements, but will improve scalability in terms of storage and programming time.
+
+## 2.4 Warm Boot Requirements
+- Unchanged - the new next hop group table must be compatible with existing warm boot requirements.
+
+# 3 Design
+## 3.1 Overview
+This design directly changes APP_DB and orchagent, along with minor changes to support the existing CLI when using the new APP_DB table.
+
+## 3.2 DB Changes
+### 3.2.1 APP DB
+Currently the ROUTE_TABLE in APP_DB includes all next hop information for the route, as follows:
+```
+### ROUTE_TABLE
+    ;Stores a list of routes
+    ;Status: Mandatory
+    key           = ROUTE_TABLE:prefix
+    nexthop       = *prefix, ;IP addresses separated “,” (empty indicates no gateway)
+    ifname        = *PORT_TABLE.key,   ; zero or more separated by “,” (zero indicates no interface)
+    blackhole     = BIT ; Set to 1 if this route is a blackhole (or null0)
+ ```
+
+This design adds a new NEXT_HOP_GROUP_TABLE, to store next hop group information to be used by one or more routes.
+```
+### NEXT_HOP_GROUP_TABLE
+    ;Stores a list of groups of one or more next hops
+    ;Status: Mandatory
+    key           = NEXT_HOP_GROUP_TABLE:string ; arbitrary identifier for the next hop group
```

Comment:
>arbitrary identifier [](start = 50, length = 20)

"arbitrary identifier" is not well defined.

Reply:
Would just "name of the next hop group" be suitable? There are no restrictions on it, so wanted to convey that.

Reply:
I guess you mean to use random string? I would like your implementation choice in this design doc.

Reply:
I've changed the description slightly, and added a further clarifying comment to the doc. It's entirely up to the programming application how the string is chosen - it could be deterministic, or random.